java/garbage-collector/src/main/java/sleeper/garbagecollector/GarbageCollector.java

```diff
@@ -79,28 +80,20 @@ public void runAtTime(Instant startTime) throws StateStoreException, IOException
             Instant deletionTime = startTime.minus(delayBeforeDeletion, ChronoUnit.MINUTES);
             Iterator<String> readyForGC = stateStore.getReadyForGCFilenamesBefore(deletionTime).iterator();
 
-            int numberDeleted = 0;
-            while (readyForGC.hasNext() && numberDeleted < garbageCollectorBatchSize) {
+            List<String> deletedFilenames = new ArrayList<>();
+            while (readyForGC.hasNext() && deletedFilenames.size() < garbageCollectorBatchSize) {
                 String filename = readyForGC.next();
-                deleteFileAndUpdateStateStore(filename, stateStore, conf);
-                numberDeleted++;
+                deleteFiles(filename, conf);
+                deletedFilenames.add(filename);
             }
-            LOGGER.info("{} files deleted for table {}", numberDeleted, tableId);
-            totalDeleted += numberDeleted;
+            stateStore.deleteReadyForGCFiles(deletedFilenames);
+            LOGGER.info("{} files deleted for table {}", deletedFilenames.size(), tableId);
+            totalDeleted += deletedFilenames.size();
```

Comment:
I think we should repeat this loop until all the files have been consumed from the iterable, i.e. consume them in batches, delete that batch, update the state store, then move on to the next batch.